File: integrations/fcpx-extension/MAMS.fxbundle/Contents/Resources/js/mams-client.js
/**
 * MAMS Client for Final Cut Pro X Extension
 * Handles communication with MAMS server
 */

class MAMSClient {
    constructor() {
        this.baseUrl = '';
        this.apiKey = '';
        this.accessToken = '';
        this.isConnected = false;
        this.settings = {};
        
        // Load configuration
        this.loadConfig();
    }

    /**
     * Load configuration from localStorage
     */
    loadConfig() {
        try {
            const config = localStorage.getItem('mams_config');
            if (config) {
                const parsed = JSON.parse(config);
                this.baseUrl = parsed.serverUrl || '';
                this.apiKey = parsed.apiKey || '';
                this.accessToken = parsed.accessToken || '';
                this.settings = parsed.settings || {};
            }
        } catch (error) {
            console.error('Error loading config:', error);
        }
    }

    /**
     * Save configuration to localStorage
     */
    saveConfig(config) {
        try {
            const currentConfig = JSON.parse(localStorage.getItem('mams_config') || '{}');
            const newConfig = { ...currentConfig, ...config };
            localStorage.setItem('mams_config', JSON.stringify(newConfig));
            
            // Update instance variables
            if (config.serverUrl) this.baseUrl = config.serverUrl;
            if (config.apiKey) this.apiKey = config.apiKey;
            if (config.accessToken) this.accessToken = config.accessToken;
            if (config.settings) this.settings = { ...this.settings, ...config.settings };
        } catch (error) {
            console.error('Error saving config:', error);
        }
    }

    /**
     * Make HTTP request to MAMS API
     */
    async request(method, endpoint, options = {}) {
        const url = `${this.baseUrl.replace(/\/$/, '')}${endpoint}`;
        
        const headers = {
            'Content-Type': 'application/json',
            'User-Agent': 'MAMS-FCPX/1.0',
            ...options.headers
        };

        // Add authentication
        if (this.apiKey) {
            headers['X-API-Key'] = this.apiKey;
        } else if (this.accessToken) {
            headers['Authorization'] = `Bearer ${this.accessToken}`;
        }

        const requestOptions = {
            method,
            headers,
            ...options
        };

        // Add body for non-GET requests
        if (method !== 'GET' && options.data) {
            requestOptions.body = JSON.stringify(options.data);
        }

        try {
            const response = await fetch(url, requestOptions);
            
            if (!response.ok) {
                const errorText = await response.text();
                throw new Error(`HTTP ${response.status}: ${errorText}`);
            }

            const contentType = response.headers.get('content-type');
            if (contentType && contentType.includes('application/json')) {
                return await response.json();
            } else {
                return await response.text();
            }
        } catch (error) {
            console.error(`Request failed: ${method} ${url}`, error);
            throw error;
        }
    }

    /**
     * Test connection to MAMS server
     */
    async testConnection() {
        try {
            await this.request('GET', '/api/v1/health');
            this.isConnected = true;
            return true;
        } catch (error) {
            this.isConnected = false;
            console.error('Connection test failed:', error);
            return false;
        }
    }

    /**
     * Login to MAMS server
     */
    async login(username, password) {
        try {
            const response = await this.request('POST', '/api/v1/auth/login', {
                data: { username, password }
            });
            
            if (response.access_token) {
                this.accessToken = response.access_token;
                this.saveConfig({ accessToken: response.access_token });
                this.isConnected = true;
                return true;
            }
            
            return false;
        } catch (error) {
            console.error('Login failed:', error);
            return false;
        }
    }

    /**
     * Logout from MAMS server
     */
    async logout() {
        try {
            if (this.accessToken) {
                await this.request('POST', '/api/v1/auth/logout');
            }
        } catch (error) {
            console.warn('Logout request failed:', error);
        } finally {
            this.accessToken = '';
            this.isConnected = false;
            this.saveConfig({ accessToken: '' });
        }
    }

    /**
     * Search for assets
     */
    async searchAssets(query = '', filters = {}, page = 1, limit = 20) {
        const params = new URLSearchParams({
            q: query,
            page: page.toString(),
            limit: limit.toString(),
            ...filters
        });

        try {
            const response = await this.request('GET', `/api/v1/assets/search?${params}`);
            return {
                assets: response.data || [],
                total: response.meta?.total || 0,
                page: response.meta?.page || 1,
                pages: response.meta?.pages || 1
            };
        } catch (error) {
            console.error('Asset search failed:', error);
            return { assets: [], total: 0, page: 1, pages: 1 };
        }
    }

    /**
     * Get asset details by ID
     */
    async getAsset(assetId) {
        try {
            const response = await this.request('GET', `/api/v1/assets/${assetId}`);
            return response.data || response;
        } catch (error) {
            console.error(`Failed to get asset ${assetId}:`, error);
            return null;
        }
    }

    /**
     * Get asset metadata
     */
    async getAssetMetadata(assetId) {
        try {
            const response = await this.request('GET', `/api/v1/assets/${assetId}/metadata`);
            return response.data || response;
        } catch (error) {
            console.error(`Failed to get metadata for asset ${assetId}:`, error);
            return {};
        }
    }

    /**
     * Download asset
     */
    async downloadAsset(assetId, quality = 'proxy') {
        try {
            const response = await this.request('GET', `/api/v1/assets/${assetId}/download`, {
                data: { quality }
            });
            
            if (response.url) {
                return {
                    url: response.url,
                    filename: response.filename,
                    size: response.size
                };
            }
            
            return null;
        } catch (error) {
            console.error(`Failed to download asset ${assetId}:`, error);
            return null;
        }
    }

    /**
     * Get asset proxy URL for streaming
     */
    async getProxyUrl(assetId) {
        try {
            const response = await this.request('GET', `/api/v1/assets/${assetId}/proxy`);
            return response.url || null;
        } catch (error) {
            console.error(`Failed to get proxy URL for asset ${assetId}:`, error);
            return null;
        }
    }

    /**
     * Upload asset to MAMS
     */
    async uploadAsset(file, metadata = {}) {
        try {
            const formData = new FormData();
            formData.append('file', file);
            formData.append('metadata', JSON.stringify(metadata));

            const headers = {};
            if (this.apiKey) {
                headers['X-API-Key'] = this.apiKey;
            } else if (this.accessToken) {
                headers['Authorization'] = `Bearer ${this.accessToken}`;
            }

            const response = await fetch(`${this.baseUrl}/api/v1/assets/upload`, {
                method: 'POST',
                headers,
                body: formData
            });

            if (!response.ok) {
                throw new Error(`Upload failed: ${response.statusText}`);
            }

            const result = await response.json();
            return result.data || result;
        } catch (error) {
            console.error('Asset upload failed:', error);
            throw error;
        }
    }

    /**
     * Update asset metadata
     */
    async updateAssetMetadata(assetId, metadata) {
        try {
            const response = await this.request('PATCH', `/api/v1/assets/${assetId}/metadata`, {
                data: metadata
            });
            return response.data || response;
        } catch (error) {
            console.error(`Failed to update metadata for asset ${assetId}:`, error);
            return false;
        }
    }

    /**
     * Create project in MAMS
     */
    async createProject(projectData) {
        try {
            const response = await this.request('POST', '/api/v1/projects', {
                data: projectData
            });
            return response.data || response;
        } catch (error) {
            console.error('Failed to create project:', error);
            return null;
        }
    }

    /**
     * Get list of projects
     */
    async getProjects(page = 1, limit = 50) {
        try {
            const params = new URLSearchParams({
                page: page.toString(),
                limit: limit.toString()
            });

            const response = await this.request('GET', `/api/v1/projects?${params}`);
            return {
                projects: response.data || [],
                total: response.meta?.total || 0,
                page: response.meta?.page || 1,
                pages: response.meta?.pages || 1
            };
        } catch (error) {
            console.error('Failed to get projects:', error);
            return { projects: [], total: 0, page: 1, pages: 1 };
        }
    }

    /**
     * Sync project with MAMS
     */
    async syncProject(projectData) {
        try {
            const response = await this.request('POST', '/api/v1/projects/sync', {
                data: projectData
            });
            return response.success || response.status === 'success';
        } catch (error) {
            console.error('Failed to sync project:', error);
            return false;
        }
    }

    /**
     * Get collections
     */
    async getCollections() {
        try {
            const response = await this.request('GET', '/api/v1/collections');
            return response.data || [];
        } catch (error) {
            console.error('Failed to get collections:', error);
            return [];
        }
    }

    /**
     * Get keywords/tags
     */
    async getKeywords() {
        try {
            const response = await this.request('GET', '/api/v1/keywords');
            return response.data || [];
        } catch (error) {
            console.error('Failed to get keywords:', error);
            return [];
        }
    }

    /**
     * Download LUT file
     */
    async downloadLUT(lutName) {
        try {
            const response = await this.request('GET', `/api/v1/luts/${lutName}/download`);
            return response.url || null;
        } catch (error) {
            console.error(`Failed to download LUT ${lutName}:`, error);
            return null;
        }
    }

    /**
     * Get user profile
     */
    async getUserProfile() {
        try {
            const response = await this.request('GET', '/api/v1/auth/profile');
            return response.data || response;
        } catch (error) {
            console.error('Failed to get user profile:', error);
            return null;
        }
    }

    /**
     * Create event in MAMS
     */
    async createEvent(eventData) {
        try {
            const response = await this.request('POST', '/api/v1/events', {
                data: eventData
            });
            return response.data || response;
        } catch (error) {
            console.error('Failed to create event:', error);
            return null;
        }
    }

    /**
     * Export timeline to MAMS
     */
    async exportTimeline(timelineData) {
        try {
            const response = await this.request('POST', '/api/v1/timelines', {
                data: timelineData
            });
            return response.data || response;
        } catch (error) {
            console.error('Failed to export timeline:', error);
            return null;
        }
    }

    /**
     * Get server info
     */
    async getServerInfo() {
        try {
            const response = await this.request('GET', '/api/v1/info');
            return response.data || response;
        } catch (error) {
            console.error('Failed to get server info:', error);
            return null;
        }
    }

    /**
     * Format file size
     */
    formatFileSize(bytes) {
        if (bytes === 0) return '0 Bytes';
        
        const k = 1024;
        const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB'];
        const i = Math.floor(Math.log(bytes) / Math.log(k));
        
        return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
    }

    /**
     * Format duration from seconds
     */
    formatDuration(seconds) {
        if (!seconds || seconds <= 0) return '00:00:00';
        
        const hours = Math.floor(seconds / 3600);
        const minutes = Math.floor((seconds % 3600) / 60);
        const secs = Math.floor(seconds % 60);
        
        return [hours, minutes, secs]
            .map(v => v.toString().padStart(2, '0'))
            .join(':');
    }

    /**
     * Parse timecode to seconds
     */
    parseTimecode(timecode) {
        if (!timecode) return 0;
        
        const parts = timecode.split(':').map(Number);
        if (parts.length === 3) {
            return parts[0] * 3600 + parts[1] * 60 + parts[2];
        } else if (parts.length === 4) {
            // HH:MM:SS:FF format - ignore frames for now
            return parts[0] * 3600 + parts[1] * 60 + parts[2];
        }
        
        return 0;
    }

    /**
     * Get asset type icon
     */
    getAssetTypeIcon(type) {
        const icons = {
            video: 'icon-play',
            audio: 'icon-volume',
            image: 'icon-image',
            project: 'icon-folder',
            sequence: 'icon-film',
            timeline: 'icon-timeline'
        };
        
        return icons[type] || 'icon-file';
    }

    /**
     * Validate configuration
     */
    validateConfig() {
        const errors = [];
        
        if (!this.baseUrl) {
            errors.push('Server URL is required');
        } else {
            try {
                new URL(this.baseUrl);
            } catch {
                errors.push('Invalid server URL format');
            }
        }
        
        if (!this.apiKey && !this.accessToken) {
            errors.push('API key or access token is required');
        }
        
        return errors;
    }

    /**
     * Get connection status
     */
    getConnectionStatus() {
        return {
            connected: this.isConnected,
            serverUrl: this.baseUrl,
            hasAuth: !!(this.apiKey || this.accessToken)
        };
    }
}

// Export as global for use in extension
window.MAMSClient = MAMSClient;

File: integrations/davinci-resolve/utils/mams_client.py
"""
MAMS Client for DaVinci Resolve Integration
Handles communication with MAMS server
"""

import json
import requests
import os
from datetime import datetime
from typing import Dict, List, Optional, Any
import tempfile
import shutil

class MAMSClient:
    def __init__(self):
        self.base_url = ""
        self.api_key = ""
        self.access_token = ""
        self.session = requests.Session()
        self._load_config()
        
    def _load_config(self):
        """Load configuration from settings file"""
        config_path = os.path.expanduser("~/.mams/resolve_config.json")
        if os.path.exists(config_path):
            try:
                with open(config_path, 'r') as f:
                    config = json.load(f)
                    self.base_url = config.get('server_url', '')
                    self.api_key = config.get('api_key', '')
                    self.access_token = config.get('access_token', '')
            except Exception as e:
                print(f"Error loading config: {e}")
    
    def save_config(self, server_url: str, api_key: str):
        """Save configuration to settings file"""
        config_dir = os.path.expanduser("~/.mams")
        os.makedirs(config_dir, exist_ok=True)
        
        config = {
            'server_url': server_url,
            'api_key': api_key,
            'access_token': self.access_token
        }
        
        config_path = os.path.join(config_dir, 'resolve_config.json')
        with open(config_path, 'w') as f:
            json.dump(config, f, indent=2)
        
        self.base_url = server_url
        self.api_key = api_key
    
    def _request(self, method: str, endpoint: str, **kwargs) -> requests.Response:
        """Make HTTP request to MAMS API"""
        url = f"{self.base_url.rstrip('/')}{endpoint}"
        
        headers = {
            'Content-Type': 'application/json',
            'User-Agent': 'MAMS-Resolve/1.0'
        }
        
        if self.api_key:
            headers['X-API-Key'] = self.api_key
        elif self.access_token:
            headers['Authorization'] = f'Bearer {self.access_token}'
        
        if 'headers' in kwargs:
            headers.update(kwargs['headers'])
        kwargs['headers'] = headers
        
        response = self.session.request(method, url, **kwargs)
        response.raise_for_status()
        return response
    
    def test_connection(self) -> bool:
        """Test connection to MAMS server"""
        try:
            response = self._request('GET', '/api/v1/health')
            return response.status_code == 200
        except:
            return False
    
    def login(self, username: str, password: str) -> bool:
        """Login to MAMS server"""
        try:
            response = self._request('POST', '/api/v1/auth/login', 
                                   json={'username': username, 'password': password})
            data = response.json()
            self.access_token = data.get('access_token', '')
            return bool(self.access_token)
        except:
            return False
    
    def search_assets(self, query: str = "", filters: Dict[str, Any] = None) -> List[Dict]:
        """Search for assets in MAMS"""
        params = {'q': query}
        if filters:
            params.update(filters)
        
        try:
            response = self._request('GET', '/api/v1/assets/search', params=params)
            data = response.json()
            return data.get('data', [])
        except Exception as e:
            print(f"Search error: {e}")
            return []
    
    def get_asset(self, asset_id: str) -> Optional[Dict]:
        """Get asset details by ID"""
        try:
            response = self._request('GET', f'/api/v1/assets/{asset_id}')
            return response.json()
        except:
            return None
    
    def get_asset_metadata(self, asset_id: str) -> Dict[str, Any]:
        """Get asset metadata"""
        try:
            response = self._request('GET', f'/api/v1/assets/{asset_id}/metadata')
            return response.json()
        except:
            return {}
    
    def download_asset(self, asset_id: str, quality: str = 'proxy') -> Optional[str]:
        """Download asset to local temporary file"""
        try:
            # Get download URL
            response = self._request('GET', f'/api/v1/assets/{asset_id}/download', 
                                   params={'quality': quality})
            download_data = response.json()
            download_url = download_data.get('url')
            filename = download_data.get('filename', f'asset_{asset_id}')
            
            if not download_url:
                return None
            
            # Download file
            file_response = requests.get(download_url, stream=True)
            file_response.raise_for_status()
            
            # Save to temporary file
            temp_dir = tempfile.mkdtemp(prefix='mams_resolve_')
            local_path = os.path.join(temp_dir, filename)
            
            with open(local_path, 'wb') as f:
                shutil.copyfileobj(file_response.raw, f)
            
            return local_path
            
        except Exception as e:
            print(f"Download error: {e}")
            return None
    
    def get_proxy_url(self, asset_id: str) -> Optional[str]:
        """Get proxy URL for streaming"""
        try:
            response = self._request('GET', f'/api/v1/assets/{asset_id}/proxy')
            data = response.json()
            return data.get('url')
        except:
            return None
    
    def upload_asset(self, file_path: str, metadata: Dict[str, Any] = None) -> Optional[str]:
        """Upload asset to MAMS"""
        try:
            with open(file_path, 'rb') as f:
                files = {'file': f}
                data = {'metadata': json.dumps(metadata or {})}
                
                headers = {}
                if self.api_key:
                    headers['X-API-Key'] = self.api_key
                elif self.access_token:
                    headers['Authorization'] = f'Bearer {self.access_token}'
                
                response = requests.post(
                    f"{self.base_url}/api/v1/assets/upload",
                    files=files,
                    data=data,
                    headers=headers
                )
                response.raise_for_status()
                
                result = response.json()
                return result.get('id')
                
        except Exception as e:
            print(f"Upload error: {e}")
            return None
    
    def create_project(self, project_data: Dict[str, Any]) -> Optional[str]:
        """Create project in MAMS"""
        try:
            response = self._request('POST', '/api/v1/projects', json=project_data)
            data = response.json()
            return data.get('id')
        except Exception as e:
            print(f"Project creation error: {e}")
            return None
    
    def sync_project(self, project_data: Dict[str, Any]) -> bool:
        """Sync project with MAMS"""
        try:
            response = self._request('POST', '/api/v1/projects/sync', json=project_data)
            return response.status_code == 200
        except Exception as e:
            print(f"Project sync error: {e}")
            return False
    
    def get_projects(self) -> List[Dict]:
        """Get list of projects"""
        try:
            response = self._request('GET', '/api/v1/projects')
            data = response.json()
            return data.get('data', [])
        except:
            return []
    
    def get_collections(self) -> List[Dict]:
        """Get list of collections"""
        try:
            response = self._request('GET', '/api/v1/collections')
            data = response.json()
            return data.get('data', [])
        except:
            return []
    
    def download_lut(self, lut_name: str) -> Optional[str]:
        """Download LUT file for color grading"""
        try:
            response = self._request('GET', f'/api/v1/luts/{lut_name}/download')
            download_data = response.json()
            download_url = download_data.get('url')
            
            if not download_url:
                return None
            
            # Download LUT file
            lut_response = requests.get(download_url)
            lut_response.raise_for_status()
            
            # Save to temporary file
            temp_dir = tempfile.mkdtemp(prefix='mams_luts_')
            lut_path = os.path.join(temp_dir, f'{lut_name}.cube')
            
            with open(lut_path, 'wb') as f:
                f.write(lut_response.content)
            
            return lut_path
            
        except Exception as e:
            print(f"LUT download error: {e}")
            return None
    
    def update_asset_metadata(self, asset_id: str, metadata: Dict[str, Any]) -> bool:
        """Update asset metadata"""
        try:
            response = self._request('PATCH', f'/api/v1/assets/{asset_id}/metadata', 
                                   json=metadata)
            return response.status_code == 200
        except Exception as e:
            print(f"Metadata update error: {e}")
            return False

# Global instance
mams_client = MAMSClient()